### DISCUSSION TITLE
Reusing a refresh token within a configurable reuse interval

### DISCUSSION THREAD

**Discussion by adanielyan** [directus/directus#17947] [Feature Requests]
### Summary

I recently ran into an issue with my app trying to refresh a directus access token through API. I use Remix as my app framework which uses concurrent requests to fetch data from directus API (concurrency is by Remix design). When I call an API that needs authentication I run into problems if the token has to be refreshed. The problem is that multiple loaders may try to refresh the access token at the same time. These concurrent requests may happen within a few milliseconds. However once the refresh token is used to retrieve a new access token it cannot be reused, which causes the subsequent requests with the same refresh token to fail. 

I recently learned that Supabase has this problem solved by enabling a refresh token reuse grace period, that can be configured with REFRESH_TOKEN_REUSE_INTERVAL variable. How difficult will it be to implement something similar in Directus? 

https://supabase.com/docs/reference/javascript/auth-setsession

### Motivation

See above

### Detailed Design

Reusing refresh token within a configurable interval if enabled through env vars, such as ENABLE_REFRESH_TOKEN_REUSE and REFRESH_TOKEN_REUSE_INTERVAL. Alternativly we can only use the REFRESH_TOKEN_REUSE_INTERVAL var and have it set to 0 by default. 

When a refresh token is used first time a timestamp of its reuse expiration is stored alongside with it in the database. If another attempt is made to use the same refresh token, the system will check if the current timestamp is smaller than the one stored in the database for that refresh token, and if so a new token pair is generated. Otherwise the refresh token is considered expired and can be removed from db. 

### Must Have:
- An env var to set the refresh token reuse interval (0 by default)
- Ability to store multiple refresh tokens per user, along with their expiration dates and reuse interval timestamps.

### Should Have:
- ?

### Could Have:
- A separate env var for enabling/disabling this feature

### Won't Have:
- ?

### Drawbacks

Very low security risk, although in practice it will not have any significant impact. 

#17888 may pose a higher security risk. 

### Alternatives

None that I am aware of

### Adoption Strategy

TBD

**Comment by rijkvanzanten** [comment]
> How difficult will it be to implement something similar in Directus?

I suppose we'd have to update the logic that currently removes the session token on refresh and replace it with some logic that instead updates the expiration date to NOW + configured interval 🤔

**Comment by faizan47** [comment]
I am facing the same issue. @adanielyan Were you able to find a solution? Thanks.

**Comment by CalebBarnes** [comment]
Super difficult to use directus auth refresh with nextauth because the refresh tokens are one time use. 

I think it would be same issue for anything that makes multiple requests simultaneously that require token refresh.

**Reply by rijkvanzanten** [reply]
IIRC the SDK already handles that gracefully by awaiting the refresh in every other call that's being made simultaneously, so you don't end up making duplicate /refresh requests 👍🏻

**Reply by adanielyan** [reply]
Only if used in the client (browser). But if used on the server that serves different authenticated users it can't handle it gracefully.

**Reply by adanielyan** [reply]
But even in the client, if multiple tabs with the app are open, there can be a racing condition.

The only bulletproof solution to this problem is supporting a grace period for used refresh tokens.

**Reply by CalebBarnes** [reply]
The SDK is not an option when using nextauth server side. Nextjs server components + nextauth is also like 10x faster than using the SDK. (Caching + request deduping)

I am currently using Redis to store users refresh tokens and locking the item to prevent race conditons. 

If we could reuse the refresh token for just a few additional seconds I could wipe out the need for this extra redis db for each project. Along with reducing the complexity for refreshing auth tokens.

**Reply by connorwinston** [reply]
@br41nslug Does this feature request fall into work on the new SDK rewrite?

**Reply by adanielyan** [reply]
I don't think it can be fixed in the SDK without enabling this feature in the directus itself first.

**Reply by CalebBarnes** [reply]
This has nothing to do with SDK. It needs to be enabled in Directus. This will happen for any refresh requests, and a lot of people aren't even using the SDK.

Theoretically this change shouldn't affect the SDK at all.

**Reply by CalebBarnes** [reply]
Even if it was like a 30 second grace period it would fix everything.

**Comment by nles** [comment]
Facing similar issue with a Next.js project using the app-router. I'm using Directus SDK in the frontend, and things are fine there, refreshing the token works automatically and easily. I'm also sharing the access token with the server to get server components working, and I'm using a middleware to handle the token refresh process (pretty much [this](https://www.youtube.com/watch?v=Bh1TOpOcGJQ&t=775s) adopted to work with Directus tooling).

When the session expires, the middleware will try to refresh the token, and often it works just fine. Sometimes tho, there are situations where two routes get triggered simultaneously, causing the middleware to run twice, with the effect that only the first refresh process can succeed, and the other one will fail. 

Small reuse interval would fix this and help get rid of some complexity I have to add to manage these cases where two requests are trigged simultaneously...

**Comment by adanielyan** [comment]
> Small reuse interval would fix this and help get rid of some complexity I have to add to manage these cases where two requests are trigged simultaneously...

Indeed, this is the only reasonable way to implement token refresh without having an issue of racing conditions.

**Reply by CalebBarnes** [reply]
Yeah it's kind of crazy that I need to have a Redis server up storing the refresh token and locking it with redlock just to prevent race conditions right now. This change would be so small and have a huge impact.

**Reply by adanielyan** [reply]
I believe this change will impact the db structure and therefore is not that small. But it is true it will have a big positive impact on DX.

**Comment by A4TIF** [comment]
Looking for a solution to this as well. Next-Auth with credentials is just not usable anymore without reusable old refresh tokens because of the concurrent requests.

**Reply by CalebBarnes** [reply]
Right now you can use redis with redlock to prevent race conditions by storing and accessing the refresh tokens there. Hopefully this won't be need in the future.

**Reply by CalebBarnes** [reply]
Using ioredis and redlock

```js
import Redis from "ioredis";
import Redlock from "redlock";

const redis = new Redis(process.env.REDIS_URL as string);

const redlock = new Redlock([redis], {
  driftFactor: 0.01,
  retryCount: 10,
  retryDelay: 200,
  retryJitter: 200,
  automaticExtensionThreshold: 500,
});
```

Here is an example of my jwt callback. 

On first login we are just storing the user's token in redis, then if the token is expired and needs to be refreshed we lock the item in the first instance this is called. all other calls that happen at same time will await that lock and then use the updated token that got stored in redis after the refresh.

```js
    async jwt({ token, user, account }) {
      if (account && user) {
        token = {
          ...token,
          accessToken: user?.data?.access_token,
          expires: Date.now() + user?.data?.expires,
          refreshToken: user?.data?.refresh_token,
          error: user?.data?.error,
          userId: user?.id,
        };

        await redis.set(`token:${token.userId}`, JSON.stringify(token));

        return token;
      }

      if (Date.now() < token.expires) {
        return token;
      }

      return await redlock.using(
        [token.userId, "jwt-refresh"],
        5000,
        async () => {
          const redisToken = await redis.get(`token:${token.userId}`);
          const currentToken = redisToken && JSON.parse(redisToken ?? "");

          if (Date.now() < currentToken.expires) {
            return currentToken;
          }

          const newToken = await refreshAccessToken(currentToken);
          await redis.set(`token:${token.userId}`, JSON.stringify(newToken));
          return newToken;
        }
      );
    }
    
```
    
and my refreshAccessToken function is just making the request
```js
async function refreshAccessToken(token) {
  try {
    const response = await fetch(
      `${process.env.NEXT_PUBLIC_DIRECTUS_ENDPOINT}/auth/refresh`,
      {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          refresh_token: token.refreshToken,
        }),
      }
    );

    const refreshedTokens = await response.json();

    if (response.ok && refreshedTokens) {
      return {
        ...token,
        accessToken: refreshedTokens?.data?.access_token,
        expires: Date.now() + refreshedTokens?.data?.expires,
        refreshToken: refreshedTokens?.data?.refresh_token,
      };
    }
  } catch (error) {
    console.log(
      new Date().toUTCString() + " Error in refreshAccessToken:",
      error
    );

    return {
      ...token,
      error: "RefreshAccessTokenError",
    };
  }
}
```

**Reply by A4TIF** [reply]
Thanks! This helps, will test it out shortly.

**Reply by CalebBarnes** [reply]
👍 Next auth is really nice with Directus once you get it working.
It's kind of annoying having to have this redis db set up for each directus project so I just created a docker compose file that has one in there.

**Reply by A4TIF** [reply]
Tested and worked like a charm. Yeah a configurable reuse interval feature will definitely help, as running redis feels a bit too much at the moment. It was my first time working on next-auth, and was almost about to jump the ship to supabase (although not an easy one to self-host with docker)! Day saved, thanks to you as Directus is very straightforward and easy to work with. Thanks.

**Comment by zakwear** [comment]
Redis is worth it for next auth. Supports multi-tab browsing, refresh
works, and Redis doesn’t care about how many separate apps/dbs it supports.

On Sun, Jul 30, 2023 at 10:16 AM Aatif Khan ***@***.***>
wrote:

> Tested and worked like a charm. Yeah a configurable reuse interval feature
> will definitely help, as running redis feels a bit too much at the moment.
> It was my first time working on next-auth, and was almost about to jump the
> ship to supabase (although not an easy one to self-host with docker)! Day
> saved, thanks to you as Directus is very straightforward and easy to work
> with. Thanks.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/directus/directus/discussions/17947#discussioncomment-6588503>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANHY7LV3G5G3NXI44RYW3OLXS2JH5ANCNFSM6AAAAAAWIK2CKE>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>
-- 
Zak Wear
Founder & CEO
Potluck Action
***@***.***
www.potluckaction.com

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
